File: 02.CSharpPartTwo/06_StringsAndTextProcessing/StringsAndTextProcessing/05ParseTags/ParseTags.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

/*Problem 5. Parse tags

You are given a text. Write a program that changes the text in all regions surrounded by the tags <upcase> and </upcase> to upper-case.
The tags cannot be nested.
 */

namespace _05ParseTags
{
    class ParseTags
    {
        static void Main(string[] args)
        {
            //For concole input:
            //Console.Write("Enter the string: ");
            //string inputString = Console.ReadLine();

            string inputString = "We are living in a <upcase>yellow submarine</upcase>. We don't have <upcase>anything</upcase> else.";

            string[] tags = { "<upcase>", "</upcase>" };

            string[] splittedInputString = inputString.Split(tags, StringSplitOptions.None);

            string subString = string.Empty;

            for (int i = 1; i < splittedInputString.GetLength(0); i += 2)
            {
                subString = splittedInputString[i].ToUpper();

                splittedInputString[i] = subString;
            }

            Console.WriteLine(string.Join("", splittedInputString));
        }
    }
}

File: 06_JavaScript/09_Strings/Strings/04.ParseTags/ParseTags.js
var example = 'We are <mixcase>living</mixcase> in a <upcase>yellow submarine</upcase>. We <mixcase>don\'t</mixcase> have <lowcase>anything</lowcase> else.';

function getTagName(text, startIndex) {
    var j,
        len = text.length,
        result = '';

    for (j = startIndex + 1; j < len; j += 1) {
        if (0 != text[j].localeCompare('>')) {
            result += text[j];
        } else {
            return result;
        }
    }

    return result;
}

function textFromTag(substring, text, start) {
    var j,
        len = text.length,
        tag = '',
        result = '';

    if (0 === substring[0].localeCompare('/')) {
        return result;
    }

    switch (substring) {
        case 'upcase':
            for (j = start + substring.length + 2; j < len; j += 1) {
                if (0 === text[j].localeCompare('>')) {
                    break;
                } else if (0 === text[j].localeCompare('<')) {
                    if (0 === text[j + 1].localeCompare('/')) {
                        return result
                    } else {
                        tag = getTagName(text, j);
                        result += textFromTag(tag, text, j);
                    }
                } else {
                    result += text[j].toUpperCase();
                }
            }
            break;
        case 'lowcase':
            for (j = start + substring.length + 2; j < len; j += 1) {
                if (0 === text[j].localeCompare('>')) {
                    break;
                } else if (0 === text[j].localeCompare('<')) {
                    if (0 === text[j + 1].localeCompare('/')) {
                        return result
                    } else {
                        tag = getTagName(text, j);
                        result += textFromTag(tag, text, j);
                    }
                } else {
                    result += text[j].toLowerCase();
                }
            }
            break;
        case 'mixcase':
            for (j = start + substring.length + 2; j < len; j += 1) {
                if (0 === text[j].localeCompare('>')) {
                    break;
                } else if (0 === text[j].localeCompare('<')) {
                    if (0 === text[j + 1].localeCompare('/')) {
                        return result
                    } else {
                        tag = getTagName(text, j);
                        result += textFromTag(tag, text, j);
                    }
                } else {
                    var random = Math.random() * 10;

                    if (random > 5) {
                        result += text[j].toUpperCase();
                    } else {
                        result += text[j].toLowerCase();
                    }
                }
            }
            break;
    }

    return result;
}

function parseTags(input) {
    var i,
        len = input.length,
        substring = '',
        result = '';

    for (i = 0; i < len; i += 1) {
        if (0 === input[i].localeCompare('<')) {
            substring = getTagName(input, i);

            var subResult = textFromTag(substring, input, i)

            i += (substring.length * 2) + 4;
            i += subResult.length;

            result += subResult;
        } else {
            result += input[i];
        }
    }
    return result;
}

console.log(parseTags(example));